frame/treasury/src/lib.rs

```diff
@@ -625,6 +786,333 @@ decl_module! {
+				let mut bounty = maybe_bounty.as_mut().ok_or(Error::<T>::InvalidIndex)?;
+				match &bounty.status {
+					BountyStatus::Active {
+						curator,
+						..
+					} => {
+						ensure!(signer == *curator, Error::<T>::RequireCurator);
+					},
+					_ => return Err(Error::<T>::UnexpectedStatus.into()),
+				}
+				bounty.status = BountyStatus::PendingPayout {
+					curator: signer,
+					beneficiary: beneficiary.clone(),
+					unlock_at: system::Module::<T>::block_number() + T::BountyDepositPayoutDelay::get(),
+				};
+
+				Ok(())
+			})?;
+
+			Self::deposit_event(Event::<T>::BountyAwarded(bounty_id, beneficiary));
+		}
+
+		/// Claim the payout from an awarded bounty after payout delay.
+		///
+		/// The dispatch origin for this call must be the beneficiary of this bounty.
+		///
+		/// - `bounty_id`: Bounty ID to claim.
+		#[weight = T::WeightInfo::claim_bounty()]
+		fn claim_bounty(origin, #[compact] bounty_id: BountyIndex) {
+			let _ = ensure_signed(origin)?; // anyone can trigger claim
+
+			Bounties::<T>::try_mutate_exists(bounty_id, |maybe_bounty| -> DispatchResult {
+				let bounty = maybe_bounty.take().ok_or(Error::<T>::InvalidIndex)?;
+				if let BountyStatus::PendingPayout { curator, beneficiary, unlock_at } = bounty.status {
+					ensure!(system::Module::<T>::block_number() >= unlock_at, Error::<T>::Premature);
+					let bounty_account = Self::bounty_account_id(bounty_id);
+					let balance = T::Currency::free_balance(&bounty_account);
+					let fee = bounty.fee;
+					let payout = balance.saturating_sub(fee);
+					let _ = T::Currency::unreserve(&curator, bounty.curator_deposit);
+					let _ = T::Currency::transfer(&bounty_account, &curator, fee, AllowDeath); // should not fail
+					let _ = T::Currency::transfer(&bounty_account, &beneficiary, payout, AllowDeath); // should not fail
+					*maybe_bounty = None;
+
+					BountyDescriptions::remove(bounty_id);
+
+					Self::deposit_event(Event::<T>::BountyClaimed(bounty_id, payout, beneficiary));
+					Ok(())
+				} else {
+					Err(Error::<T>::UnexpectedStatus.into())
+				}
+			})?;
+		}
+
+		/// Cancel an active bounty. All the funds will be send to treasury.
+		///
+		/// Only curator is able to cancel active bounty.
+		/// Otherwise `T::RejectOrigin` is able to cancel funded bounty.
+		///
+		/// - `bounty_id`: Bounty ID to cancel.
+		#[weight = T::WeightInfo::cancel_bounty()]
+		fn cancel_bounty(origin, #[compact] bounty_id: BountyIndex) {
+			Bounties::<T>::try_mutate_exists(bounty_id, |maybe_bounty| -> DispatchResult {
+				let bounty = maybe_bounty.as_ref().ok_or(Error::<T>::InvalidIndex)?;
+
+				match &bounty.status {
+					BountyStatus::Funded |
+					BountyStatus::CuratorAssigned { .. } => {
+						T::RejectOrigin::ensure_origin(origin)?;
+					},
+					BountyStatus::PendingPayout { curator, .. } => {
+						T::RejectOrigin::ensure_origin(origin)?;
+
+						let _ = T::Currency::unreserve(curator, bounty.curator_deposit);
```

Comment:
Should be slash? Would be strange to cancel a bounty when `PendingPayout` for any other reason?

Reply:
I can't think much reason but I want to be consistent so cancel means no slash and unassign means slash.

Reply:
This is not a really valid argument afact.

A curator has marked a bounty as ready for payout. The payout is pending. The council (Reject Origin) wants to cancel the bounty now... why?

Either should not be allowed to be canceled since everything is green and complete, or should be a slash to the curator because the council is intervening to some malicious action.